File: rust/tests/nginx.rs

use std::io::{Read, Write};
use std::net::TcpStream;

use disposables::container::ContainerParams;
use disposables::protocol::V1Event;
use disposables::util::try_use;


#[test]
fn normal_server() {
    drop(env_logger::try_init());

    let mut container = ContainerParams::new("docker.io/nginx:alpine")
        .port(80)
        .wait_for_port(80)
        .create().unwrap();

    assert!(matches!(container.wait(), Ok(V1Event::Ready)),
        "Container start failed, logs: {}", container.logs().unwrap());

    let mut conn = try_use(container.port(80).unwrap(), TcpStream::connect).unwrap();
    write!(conn, "GET / HTTP/1.0\nHost: localhost\n\n").unwrap();

    let mut response_buf = Vec::<u8>::new();
    conn.read_to_end(&mut response_buf).unwrap();
    let response = String::from_utf8(response_buf).unwrap();
    assert_eq!(response.split("\r\n").next().unwrap(), "HTTP/1.1 200 OK",
        "Unexpected response: {response}");
}

#[test]
fn custom_file() {
    drop(env_logger::try_init());

    let mut container = ContainerParams::new("docker.io/nginx:alpine")
        .port(80)
        .wait_for_port(80)
        .file("/usr/share/nginx/html/test.html",
            "<html></html>")
        .create().unwrap();

    let event = container.wait();
    assert!(matches!(event, Ok(V1Event::Ready)),
        "Container start failed: {event:?}, logs: {}", container.logs().unwrap());

    let mut conn = try_use(container.port(80).unwrap(), TcpStream::connect).unwrap();
    write!(conn, "GET /test.html HTTP/1.0\nHost: localhost\n\n").unwrap();

    let mut response_buf = Vec::<u8>::new();
    conn.read_to_end(&mut response_buf).unwrap();
    let response = String::from_utf8(response_buf).unwrap();
    assert_eq!(response.split("\r\n").next().unwrap(), "HTTP/1.1 200 OK",
        "Unexpected response: {response}");
}

//TODO: Delayed startup

File: java/src/test/java/io/p01def/disposables/Nginx.java

package io.p01def.disposables;

import java.net.HttpURLConnection;
import java.net.URL;

import org.junit.jupiter.api.Test;
import static org.junit.jupiter.api.Assertions.*;

import io.p01def.disposables.Container.MappedPort;
import io.p01def.disposables.protocol.V1Event;

public class Nginx {

	@Test
	public void normalServer() throws Exception {
		//Use 'docker.io/nginx:alpine' image
		Container c = new ContainerParams("docker.io/nginx:alpine")
			.port(80) //< Forward port 80
			.waitForPort(80) //< When port 80 is connectable, the container is ready
			.file("/usr/share/nginx/html/custom_file.html",
				"<html></html>".getBytes()) //< Add this file
			.create();

		V1Event event = c.waitForEvent();
		if (! (event instanceof V1Event.Ready)) {
			throw new Exception("Container start failed: " + event 
					+ ", logs: " + c.logs());
		}

		boolean success = false;
		for (MappedPort p : c.port(80)) {
			try {
				URL url = new URL("http://" + p + "/custom_file.html");
				System.out.println("Connecting to " + url);
				HttpURLConnection conn = (HttpURLConnection) url.openConnection();
				conn.setRequestMethod("GET");
				int responseCode = conn.getResponseCode();
				if (responseCode != 200) {
					throw new Exception("Unexpected response code: " + responseCode);
				}
				success = true;
				break;
			} catch (Exception e) {
				e.printStackTrace();
			}
		}
		assertTrue(success, "Cannot connect to port");
	}
}


